File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string a) {
          std::set<int> st;
          int count = 0, maxE = 0;
          int low = 0;
          for(int i = 0; i < a.length(); ++i){
             if(st.count(a[i]) == 0){
                st.insert(a[i]);
                count++;
             }
             else{
                 maxE = std::max(maxE, count);
                 while(low < a.length()){
                     if(a[low] == a[i]){
                         ++low;
                         break;
                     }
                     st.erase(a[low]);
                     ++low, --count;
                 }
               }
             maxE = std::max(maxE, count);
          }
        return maxE;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        HashMap<Character, Integer> map = new HashMap<>();
        int answer = 0;
        int low = 0;

        for (int i = 0; i < len; ++i) {
            char currentChar = s.charAt(i);
            if (map.containsKey(currentChar)) {
                low = Math.max(low, map.get(currentChar) + 1);
            }
            answer = Math.max(answer, i - low + 1);
            map.put(currentChar, i);
        }

        return answer;
    }
}